packages/opencensus-instrumentation-grpc/src/grpc.ts

```diff
@@ -0,0 +1,478 @@
+   */
+  private serverStreamAndBidiHandler(
+      plugin: GrpcPlugin, rootSpan: types.RootSpan, call: ServerCallWithMeta,
+      // tslint:disable-next-line:no-any
+      original: grpcModule.handleCall<any, any>,
+      // tslint:disable-next-line:no-any
+      self: grpcModule.handleCall<any, any>) {
+    let spanEnded = false;
+    const endSpan = () => {
+      if (!spanEnded) {
+        spanEnded = true;
+        rootSpan.end();
+      }
+    };
+
+    plugin.tracer.wrapEmitter(call);
+    call.on('finish', () => {
+      rootSpan.status = plugin.traceStatus(call.status.code);
+      rootSpan.addAttribute(
+          GrpcPlugin.ATTRIBUTE_GRPC_STATUS_CODE, call.status.code.toString());
+      // if there is an error, span is ended on error event, otherwise here
+      if (call.status.code === 0) {
+        endSpan();
+      }
+    });
+
+    call.on('error', (err: grpcModule.ServiceError) => {
+      rootSpan.addAttribute(GrpcPlugin.ATTRIBUTE_GRPC_ERROR_NAME, err.name);
+      rootSpan.addAttribute(
+          GrpcPlugin.ATTRIBUTE_GRPC_ERROR_MESSAGE, err.message);
+      endSpan();
+    });
+
+    return original.call(self, call);
+  }
+
+  /**
+   * Returns a function that patches the gRPC makeClientConstructor in order
+   * to patch all client methods.
+   * @returns {Function}  The register patched function.
+   */
+  private patchClient() {
+    const plugin = this;
+    return (original: MakeClientConstructor) => {
+      plugin.logger.debug('patchClient');
+      return function makeClientConstructor(
+          // tslint:disable-next-line:no-any
+          methods: grpcModule.ServiceDefinition<any>, serviceName: string,
+          options: grpcModule.GenericClientOptions) {
+        const client = original.apply(this, arguments);
+        shimmer.massWrap(
+            client.prototype, Object.keys(methods) as never[],
+            plugin.patchAllClientsMethods());
+        return client;
+      };
+    };
+  }
+
+
+  /**
+   * This function starts a span (child or root) immediately before the
+   * client method is invoked, and ends it either in a callback or stream
+   * event handler, depending on the method type.
+   */
+  private patchAllClientsMethods() {
+    const plugin = this;
+    // tslint:disable-next-line:no-any
+    return (original: any) => {
+      plugin.logger.debug('patchAllClientsMethods');
+      return function clientMethodTrace() {
+        const traceOptions = {
+          name: `grpc.${original.path.replace('/', '')}`,
+          type: 'CLIENT',
+        };
+        const args = Array.prototype.slice.call(arguments);
+        // Checks if this remote function call is part of an operation by
+        // checking if there is a current root span, if so, we create a child
+        // span. In case there is no root span, this means that the remote
+        // function call is the first operation, therefore we create a root
+        // span.
+        if (!plugin.tracer.currentRootSpan) {
```

Comment:
I am starting to not like having to special-case creating a root span for any client if there isn't already a root span. This complicates things for future plugin writers, as this makes it unclear what the role of a root span is, and _every_ RPC plugin will have to implement it this way. In addition, it masks async context propagation bugs.

I'd like to propose instead that we only start child spans in RPC calls, and add a config option `{ allowTopLevelChildSpans: boolean }` that users can turn on to achieve similar behavior. What do you think?